consensus/src/chained_bft/block_storage/sync_manager.rs

```diff
@@ -3,148 +3,115 @@
 
 use crate::{
     chained_bft::{
-        block_storage::{BlockReader, BlockStore, NeedFetchResult},
+        block_storage::{BlockReader, BlockStore},
         network::ConsensusNetworkImpl,
-        persistent_storage::PersistentStorage,
     },
     counters,
-    state_replication::StateComputer,
 };
 use consensus_types::{
-    block::{Block, ExecutedBlock},
+    block::Block,
     common::{Author, Payload},
     quorum_cert::QuorumCert,
     sync_info::SyncInfo,
 };
+use crypto::HashValue;
 use failure;
 use libra_logger::prelude::*;
 use libra_types::account_address::AccountAddress;
+use mirai_annotations::checked_precondition;
 use network::proto::BlockRetrievalStatus;
 use rand::{prelude::*, Rng};
 use std::{
     clone::Clone,
-    sync::Arc,
     time::{Duration, Instant},
 };
 use termion::color::*;
 
-/// SyncManager is responsible for fetching dependencies and 'catching up' for given qc/ledger info
-pub struct SyncManager<T> {
-    block_store: Arc<BlockStore<T>>,
-    storage: Arc<dyn PersistentStorage<T>>,
-    network: ConsensusNetworkImpl,
-    state_computer: Arc<dyn StateComputer<Payload = T>>,
+#[derive(Debug, PartialEq)]
+/// Whether we need to do block retrieval if we want to insert a Quorum Cert.
+pub enum NeedFetchResult {
+    QCRoundBeforeRoot,
+    QCAlreadyExist,
+    QCBlockExist,
+    NeedFetch,
 }
 
-/// Keeps the necessary context for `SyncMgr` to bring the missing information.
-pub struct SyncMgrContext {
-    pub highest_ledger_info: QuorumCert,
-    pub highest_quorum_cert: QuorumCert,
-    /// Preferred peer: this is typically the peer that delivered the original QC and
-    /// thus has higher chances to be able to return the information than the other
-    /// peers that signed the QC.
-    /// If no preferred peer provided, random peers from the given QC are going to be queried.
-    pub preferred_peer: Author,
-}
+impl<T: Payload> BlockStore<T> {
+    /// Check if we're far away from this ledger info and need to sync.
+    /// Returns false if we have this block in the tree or the root's round is higher than the
+    /// block.
+    pub fn need_sync_for_quorum_cert(
+        &self,
+        committed_block_id: HashValue,
+        qc: &QuorumCert,
+    ) -> bool {
+        // This precondition ensures that the check in the following lines
+        // does not result in an addition overflow.
+        checked_precondition!(self.root().round() < std::u64::MAX - 1);
```

Comment:
Why not use https://doc.rust-lang.org/std/primitive.u64.html#method.checked_add and avoid the comment? Also seems like something that might be a risk we should be thinking about more globally (how do we avoid doing wrapping arithmetic when there's a risk of overflow)

Reply:
Also, should we just ban forming a block at u64::MAX-3? Or maybe even lower than that -- say leave a buffer of 1 million blocks? There's not enough time to to form 2^64 blocks and that'd reduce the risk of incorrect code paths that could be triggered if you had 2f+1 faulty nodes and they signed a TC bringing you many rounds forward.

Reply:
@bmaurer these preconditions were added as part of the mirai annotations effort by Rashmi during her internship. I think we'd need a separate discussion about the status of the mirai annotations in order to decide about these specific ones :P